File: leetcode/_977.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution
{
public:
    vector<int> sortedSquares(vector<int> &nums)
    {
        int k = nums.size() - 1;
        vector<int> result(k + 1);
        int i = 0;
        int j = k;
        while (i <= j)
        {
            if (nums[i] * nums[i] <= nums[j] * nums[j])
            {
                result[k--] = nums[j] * nums[j];
                j--;
            }
            else
            {
                result[k--] = nums[i] * nums[i];
                i++;
            }
        }

        return result;
    }
};

File: leetcode/_977.py
def solution977(nums):
    n = len(nums)
    left, right = 0, n-1
    k = n-1
    result = [0] * n
    while left <= right:
        if nums[left]**2 < nums[right]**2:
            result[k] = nums[right]**2
            k -= 1
            right -= 1
        else:
            result[k] = nums[left]**2
            k -= 1
            left += 1
    
    return result